### PR TITLE
[FLINK-31642][network] Introduce the MemoryTierConsumerAgent

### PR DESCRIPTION
**Pull request by WencongLiu**
### Introduce the MemoryTierConsumerAgent

1. Implement the MemoryTierConsumerAgent
2. Introduce the MemoryTierConsumerAgent to TieredStorageConsumerClient
3. Enable NettyConnectionReader to notify upstream required segment id
 

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (not documented)

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/memory/MemoryTierConsumerAgent.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+            nettyConnectionReaders.add(
+                    nettyService.registerConsumer(
+                            storagePartitionId,
+                            partitionIdAndSubpartitionIds.get(storagePartitionId)));
+        }
+    }
+
+    @Override
+    public void start() {
+        // nothing to do.
```
Comment (xintongsong) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/memory/MemoryTierConsumerAgent.java:55]: minor: `noop`

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/netty/TieredStorageNettyServiceImpl.java [hunk 2]
```diff
@@ -170,18 +143,12 @@ public void setupInputChannels(
...
-
-            registeredNettyConnectionReaderAvailabilityAndPriorityHelpers
-                    .computeIfAbsent(partitionId, ignored -> new HashMap<>())
-                    .put(subpartitionId, helper);
+            registeredNettyConnectionReaders
+                    .remove(partitionId)
+                    .remove(subpartitionId)
+                    .complete(
+                            new NettyConnectionReaderImpl(
+                                    index, inputChannelProviders.get(index), helper));
```
Comment (xintongsong) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/netty/TieredStorageNettyServiceImpl.java:151]: How do you ensure that the consumer is already registered?

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/memory/MemoryTierConsumerAgent.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+
+/** The data client is used to fetch data from memory tier. */
+public class MemoryTierConsumerAgent implements TierConsumerAgent {
+    private final List<CompletableFuture<NettyConnectionReader>> nettyConnectionReaders =
+            new ArrayList<>();
+
+    public MemoryTierConsumerAgent(
+            Map<TieredStoragePartitionId, TieredStorageSubpartitionId>
```
Comment (xintongsong) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/memory/MemoryTierConsumerAgent.java:42]: This doesn't seem right. There's no guarantee that we only read one subpartition from each partition.

**Comment by WencongLiu** [comment]
@flinkbot run azure

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/storage/TieredStorageConsumerClient.java [hunk 4]
```diff
@@ -43,8 +68,25 @@ public void start() {
...
-        return Optional.empty();
+        Optional<Buffer> buffer = Optional.empty();
+        for (TierConsumerAgent tiereConsumerAgent : tierConsumerAgents) {
+            buffer =
+                    tiereConsumerAgent.getNextBuffer(
+                            subpartitionId, subpartitionNextSegmentIds[subpartitionId]);
+            if (buffer.isPresent()) {
+                break;
+            }
+        }
```
Comment (xintongsong) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/storage/TieredStorageConsumerClient.java:79]: It's weird we have to loop over all tier agents for reading each buffer. Shouldn't the storage client remember the current tier that each channel is reading from?

Comment (xintongsong) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/storage/TieredStorageConsumerClient.java:79]: Also, the argument name seems not right. How could a consumer reads data from different subpartitions?

**Review by xintongsong — COMMENTED** [review]
In addition to the two comments below, I added a fixup commit regarding `NettyConnectionReaderRegistration`. Please take a look.

**Comment by WencongLiu** [comment]
@flinkbot run azure

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/netty/TestingNettyConnectionReader.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+    @Override
+    public Optional<Buffer> readBuffer(int segmentId) {
+        return Optional.of(readBufferFunction.apply(segmentId));
+    }
+
+    /** Builder for {@link TestingNettyConnectionReader}. */
+    public static class Builder {
+
+        private Function<Integer, Buffer> readBufferFunction;
```
Comment (TanYuxin-tyx) [inline flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/netty/TestingNettyConnectionReader.java:43]: Use a default value to init the `readBufferFunction`, which can simplify the usage.

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/netty/TieredStorageConsumerClientTest.java [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+import org.junit.jupiter.api.Test;
+
+import java.util.Collections;
+import java.util.Optional;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/** Tests for {@link TieredStorageConsumerClient}. */
+public class TieredStorageConsumerClientTest {
```
Comment (TanYuxin-tyx) [inline flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/netty/TieredStorageConsumerClientTest.java:39]: This need not be public.

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/netty/TieredStorageConsumerClientTest.java [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+    private final TieredStoragePartitionId partitionId =
+            TieredStorageIdMappingUtils.convertId(new ResultPartitionID());
+
+    private final TieredStorageSubpartitionId subpartitionId = new TieredStorageSubpartitionId(0);
+
+    @Test
+    void testGetNextBufferFromMemoryTier() {
```
Comment (TanYuxin-tyx) [inline flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/netty/TieredStorageConsumerClientTest.java:47]: When testing consumer client, this test need not be aware of the tier implementation. I think we can add a `TestingTierFactory` to test this. Then rename the test method to `testGetNextBuffer`.

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/storage/TestingTierFactory.java [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+
+    private final Supplier<TierProducerAgent> tierProducerAgentSupplier;
+
+    private final BiFunction<
+                    List<Tuple2<TieredStoragePartitionId, TieredStorageSubpartitionId>>,
+                    TieredStorageNettyService,
+                    TierConsumerAgent>
+            tierConsumerAgentSupplier;
+
+    public TestingTierFactory(
```
Comment (TanYuxin-tyx) [inline flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/storage/TestingTierFactory.java:47]: LGTM overall, left only one comment, this should be private.

**Review by TanYuxin-tyx — APPROVED** [review]

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java [hunk 9]
```diff
@@ -217,6 +222,12 @@ public class SingleInputGate extends IndexedInputGate {
     // The consumer client will be null if the tiered storage is not enabled.
     @Nullable private final TieredStorageConsumerClient tieredStorageConsumerClient;
 
+    // The partition ids in tiered storage will be null if the tiered storage is not enabled.
+    @Nullable private final List<TieredStoragePartitionId> tieredStoragePartitionIds;
+
+    // The subpartition ids in tiered storage will be null if the tiered storage is not enabled.
+    @Nullable private final List<TieredStorageSubpartitionId> tieredStorageSubpartitionIds;
```
[the remark this answers is not in the document]
Comment (WencongLiu) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java]: `TieredStorageConsumerSpec ` has been added.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGateFactory.java [hunk 10]
```diff
@@ -158,9 +166,34 @@ public SingleInputGate create(
...
+            for (ShuffleDescriptor shuffleDescriptor : shuffleDescriptors) {
+                for (int subpartitionId = subpartitionIndexRange.getStartIndex();
+                        subpartitionId <= subpartitionIndexRange.getEndIndex();
+                        ++subpartitionId) {
+                    TieredStoragePartitionId storagePartitionId =
+                            TieredStorageIdMappingUtils.convertId(
+                                    shuffleDescriptor.getResultPartitionID());
+                    TieredStorageSubpartitionId storageSubpartitionId =
+                            new TieredStorageSubpartitionId(subpartitionId);
+                    tieredStoragePartitionIds.add(storagePartitionId);
```
[the remark this answers is not in the document]
Comment (WencongLiu) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGateFactory.java]: The initialization of `TieredStoragePartitionId` has been placed in the first `for-loop`.

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/netty/TestingTieredStorageNettyService.java [hunk 11]
```diff
@@ -49,9 +50,10 @@ public void registerProducer(
     }
 
     @Override
-    public NettyConnectionReader registerConsumer(
+    public CompletableFuture<NettyConnectionReader> registerConsumer(
             TieredStoragePartitionId partitionId, TieredStorageSubpartitionId subpartitionId) {
-        return registerConsumerFunction.apply(partitionId, subpartitionId);
+        return CompletableFuture.completedFuture(
+                registerConsumerFunction.apply(partitionId, subpartitionId));
```
[the remark this answers is not in the document]
Comment (WencongLiu) [reply flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/netty/TestingTieredStorageNettyService.java]: Fixed.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/io/benchmark/StreamNetworkBenchmarkEnvironment.java [hunk 12]
```diff
@@ -280,7 +281,8 @@ private IndexedInputGate createInputGateWithMetrics(
                         gateIndex,
                         gateDescriptor,
                         SingleInputGateBuilder.NO_OP_PRODUCER_CHECKER,
-                        newUnregisteredInputChannelMetrics());
+                        newUnregisteredInputChannelMetrics(),
+                        new TieredStorageNettyServiceImpl());
```
[the remark this answers is not in the document]
Comment (WencongLiu) [reply flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/io/benchmark/StreamNetworkBenchmarkEnvironment.java]: Fixed.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java [hunk 13]
```diff
@@ -813,6 +813,11 @@ private void checkPartitionRequestQueueInitialized() throws IOException {
                 "Bug: partitionRequestClient is not initialized before processing data and no error is detected.");
     }
 
+    @Override
+    public void notifyRequiredSegmentId(int segmentId) {
```
[the remark this answers is not in the document]
Comment (WencongLiu) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java]: I've add a UT for the erroneous case. More tests will be introduced in IT cases.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/netty/TieredStorageNettyServiceImpl.java [hunk 14]
```diff
@@ -78,29 +75,22 @@ public void registerProducer(
...
 
-        Supplier<InputChannel> inputChannelProvider =
-                registeredInputChannelProviders.get(partitionId).remove(subpartitionId);
-        if (registeredInputChannelProviders.get(partitionId).isEmpty()) {
-            registeredInputChannelProviders.remove(partitionId);
+        List<NettyConnectionReaderRegistration> registrations =
+                getReaderRegistration(partitionId, subpartitionId);
+        for (NettyConnectionReaderRegistration registration : registrations) {
+            Optional<CompletableFuture<NettyConnectionReader>> futureOpt =
+                    registration.trySetConsumer();
```
[the remark this answers is not in the document]
Comment (WencongLiu) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/netty/TieredStorageNettyServiceImpl.java:86]: Fixed.

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/storage/TestingTierFactory.java (hunk 8 above)
Comment (WencongLiu) [reply flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/storage/TestingTierFactory.java]: Fixed.

**Comment by WencongLiu** [comment]
Thanks for the review from @TanYuxin-tyx and @reswqa ! I've made a round of changes. 🎉

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/netty/TieredStorageConsumerClientTest.java [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+                                new TestingTierFactory.Builder()
+                                        .setTierConsumerAgentSupplier(MemoryTierConsumerAgent::new)
+                                        .build()),
+                        Collections.singletonList(
+                                new TieredStorageConsumerSpec(partitionId, subpartitionId)),
+                        nettyService);
+        Optional<Buffer> nextBuffer =
+                tieredStorageConsumerClient.getNextBuffer(partitionId, subpartitionId);
+        assertThat(nextBuffer).isPresent();
+        assertThat(nextBuffer.get()).isEqualTo(buffer);
```
Comment (reswqa) [inline flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/netty/TieredStorageConsumerClientTest.java:72]:
```suggestion
        assertThat(nextBuffer).hasValue(buffer);
```

Comment (reswqa) [reply flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/netty/TieredStorageConsumerClientTest.java:72]: `assertThat(nextBuffer).isPresent();` can be removed.

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannelTest.java [hunk 16]
```diff
@@ -100,6 +100,7 @@
 import static org.apache.flink.runtime.io.network.util.TestBufferFactory.createBuffer;
 import static org.apache.flink.runtime.state.CheckpointStorageLocationReference.getDefault;
 import static org.apache.flink.util.Preconditions.checkNotNull;
+import static org.assertj.core.api.AssertionsForClassTypes.assertThatThrownBy;
```
Comment (reswqa) [inline flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannelTest.java:103]: This import is not correct.

**Review by reswqa — APPROVED** [review]
Thanks for the update @WencongLiu! LGTM, merging % CI green,

**Pull request merged**